### ISSUE TITLE
How to connect extension to more than one app?

### ISSUE THREAD

**Issue by avocadoslab** [Shopify/cli#1998]
# Issue summary

`shopify extension create` allows extension to be created and connected to one Shopify app. In cases where you have development app and production app, how do we connect or register same extension with both apps?

`shopify extension connect` allows connecting only to previously selected development app so would love to know how to handle this case.

## Expected behavior

As a developer, I should have flexibility of connecting/registering extension to more than one app in cases where we maintain development, staging & production app.

## Actual behavior

Can't register already registered extension to new app.

**Comment by bhr** [comment]
We've solved the issue on our end as follows: 

- Deploy the extension directly using the API key as an argument on `shopify deploy`
- Run a script that copies the source code from the working directory to `dist` and exchanges environment dependent variables in the code

There's a `.env` file with each environment variable to be replaced. 

### Package.json
```json
{
  "config": {
    "API_KEY_PROD": "ABC", // replace with your API key
    "API_KEY_DEV": "DEF" // replace with your API key
  },
  "scripts": {
    "shopify:ext:prepare": "node extensions/prepare.mjs",
    "shopify:ext:cleanup": "node extensions/cleanup.mjs",
    "shopify:deploy:run": "shopify app deploy --force",
    "shopify:deploy:prod": "yarn shopify:ext:prepare prod && yarn shopify:deploy:run --api-key=$npm_package_config_API_KEY_PROD && yarn shopify:ext:cleanup",
    "shopify:deploy:dev": "yarn shopify:ext:prepare dev && yarn shopify:deploy:run --api-key=$npm_package_config_API_KEY_DEV && yarn shopify:ext:cleanup"
  },
  "devDependencies": {
    "@shopify/app": "^3.45.4",
    "@shopify/cli": "^3.45.4"
  }
}

```

### Prepare.mjs -  replace variables with run-time versions
```js
import { execSync } from 'child_process';
import { config } from 'dotenv';
import { readFileSync, writeFileSync } from 'fs';
import { dirname } from 'path';
import { fileURLToPath } from 'url';

const _dirname = dirname(fileURLToPath(import.meta.url));

const environment = process.argv[2];

if (environment === 'dev') {
  config({ path: `${_dirname}/.env.development` });
} else if (environment === 'prod') {
  config({ path: `${_dirname}/.env` });
} else {
  throw new Error(`Unknown env ${environment}`);
}

// copy all files from current directory into dist
const distPath = `${_dirname}/dist`;

const shell = execSync;
shell(`rm -rf ${distPath}`);
shell(`cp -r ${_dirname} ${distPath}`);

// read file and replace ENV Var
const filePath = `${distPath}/widget-ext/blocks/widget-embed.liquid`;
const buf = readFileSync(filePath);
const fileString = buf.toString();
const newString = fileString.replace('$_{MY_ENV_VAR}', process.env.MY_ENV_VAR);
writeFileSync(filePath, newString);
```

### Cleanup.mjs -  remove dist directory
```js
import { execSync } from 'child_process';
import { dirname } from 'path';
import { fileURLToPath } from 'url';

const _dirname = dirname(fileURLToPath(import.meta.url));

// copy all files from current directory into dist
const distPath = `${_dirname}/dist`;

const shell = execSync;
shell(`rm -rf ${distPath}`);

```

### Shopify.app.toml
```toml
extension_directories = ["extensions/dist/*"]
```

**Comment by ClaytonPassmore** [comment]
If we're learning to live with it, does that make it inactive? Think on it, @github-actions bot and let me know what you come up with.

**Comment by Arkham** [comment]
Hey folks, we've recently released `3.48.0` which includes support for connecting your source to multiple apps. Check out the documentation here:
- https://shopify.dev/docs/apps/tools/cli/configuration
- https://shopify.dev/docs/apps/tools/cli/managing-app-configuration-files

The CLI also supports creating a live version with `app deploy` and an unreleased version with `app deploy --no-release`. I'm going to close this, but feel free to open new issues. Thanks for reporting and for your patience!
